### PR TITLE
[WIP] Leverage Pandas for _repr_html_ and __repr__

### LINKED ISSUE #489: Rendering dataframe in Jupyter notebook
## Feature request

I hope that I am the one missing something here, because this feature request is going to sound like a very basic thing that I had assumed would have been implemented.

In pandas, when I create a dataframe and call on it in a Jupyter notebook cell, I will get an HTML rendering of it in the notebook, e.g.:

```python
# This is a Jupyter notebook cell
df = pd.read_csv(...)
df  # the HTML is rendered.
```

With cuDF, I get a textual representation instead:

```python
gdf = cudf.read_csv(...)
gdf  # the following text gets printed to screen: <cudf.DataFrame ncols=4 nrows=4208262 >
```

I had originally thought the gdf object would be rendered in HTML similar to pandas, but this was not apparent in my installation of cudf. If I have not missed out on something, then I'm hoping this can be implemented, so that cuDF's UX can match pandas' UX, and hence provide a seamless transition.

### PR DESCRIPTION
**Pull request by mrocklin**
0.8 release Todo schedule:

- [ ] Re-enable old print style when Pandas < 0.24.2 installed
- [ ] Revert unit tests and condition them on Pandas < 0.24.2
- [ ] Investigate use of to_arrow for printing Null
- [ ] Investigate use of extra line of output to identify when Nulls present

Fixes #489 

```python
In [1]: import pandas as pd
   ...: import cudf

In [2]: pdf = pd.util.testing.makeMixedDataFrame()
   ...: pdf
Out[2]:
     A    B     C          D
0  0.0  0.0  foo1 2009-01-01
1  1.0  1.0  foo2 2009-01-02
2  2.0  0.0  foo3 2009-01-05
3  3.0  1.0  foo4 2009-01-06
4  4.0  0.0  foo5 2009-01-07

In [3]: gdf = cudf.DataFrame.from_pandas(pdf)
   ...: gdf
Out[3]:
     A    B     C          D
0  0.0  0.0  foo1 2009-01-01
1  1.0  1.0  foo2 2009-01-02
2  2.0  0.0  foo3 2009-01-05
3  3.0  1.0  foo4 2009-01-06
4  4.0  0.0  foo5 2009-01-07

In [4]: gdf.A
Out[4]:
0    0.0
1    1.0
2    2.0
3    3.0
4    4.0
Name: A, dtype: float64

In [5]: gdf.D
Out[5]:
0   2009-01-01
1   2009-01-02
2   2009-01-05
3   2009-01-06
4   2009-01-07
Name: D, dtype: datetime64[ns]
```

image

### CODE REVIEW

**Comment by kkraus14** [comment]
> What are good fail cases that we should stress in the tests here?

A couple good starters would be nulls in integer columns and datetime64 columns (we only support milliseconds currently, Pandas supports only nanoseconds AFAIK, but we handle the conversion). We might be able to utilize Pandas 0.24 features based on ExtensionArrays once it's released to solve the nulls problem.

**Comment by mrocklin** [comment]
@kkraus14 can I ask you to assign this to someone else?  I think that it is important but I'm not sure what needs to be done in order to make `to_pandas` sufficiently robust.

**Comment by kkraus14** [comment]
@thomcom You've been poking around at this so all you 😄

File: python/cudf/dataframe/core.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+def get_renderable_pandas_dataframe(gdf):
+    n = pd.core.config.get_option("display.max_rows")
+    if len(gdf) <= n:
+        return gdf.to_pandas()
+    else:
+        return pd.concat(
+            [gdf.head(n + 1).to_pandas(),
+             gdf.tail(n + 1).to_pandas(),
+         ])  # enough head and tail to look the same, plus some extra
```
Comment (thomcom) [inline python/cudf/dataframe/core.py]: 👍

**Comment by mrocklin** [comment]
+1 to converting cudf int columns to nullable integer arrays in Pandas.

The concern on floating point is that we don't have an obvious choice on
whether or not to convert nulls in float to NaNs?  My gut opinion is that
we should.

**Comment by thomcom** [comment]
What's the case for leaving nulls in a floating point array as null? As far as I'm concerned `null \in NaN` but `Nan !\in null`. As far as I can see it,`null` means "object pointer that has never been initialized" and `NaN` means "Anything but the dtype that is supposed to go here."

In fact here's a good template: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html

`NaN` in numeric arrays, `None` or `NaN` in object arrays, `NaT` in datetimelike.

**Comment by harrism** [comment]
Changed to WIP as it seems to still be in progress.

**Comment by thomcom** [comment]
Hey @mrocklin I updated your PR to `branch-0.8`. It'll be great!

**Comment by thomcom** [comment]
rerun tests

**Comment by harrism** [comment]
Matt, you seemed to have done a `git rm cpp/thirdparty` or equivalent, which is why it's asking for a cpp-codeowners review. Please don't delete that. :)  

I assume this was a mistake.

**Comment by thomcom** [comment]
I updated the PR and broke it. :( One of the reasons this PR stalled is because of the weird changes to git submodules between 0.6 and 0.7 that caused problems with merges. I tried merging 0.8 into this PR, and yeah, I deleted `cpp/thirdparty`. Whoops! :D Let me see if I can add it back in. Thanks Mark!

**Comment by beckernick** [comment]
Bumping this given @mrocklin 's discussion of this topic yesterday. Is this done pending the merge conflicts / MRM submodule being resolved or still a WIP?

**Comment by thomcom** [comment]
I made some progress with this updating it and adding Integer Null support. There arose an issue involving string find_and_replace that slowed me down. I can return to it if we want to get it in to 0.8.

**Comment by thomcom** [comment]
Down to 8 failing tests here, I'm optimistic this will be reviewable soon.

**Comment by thomcom** [comment]
rerun tests

**Comment by thomcom** [comment]
This fails tests because it is only compatible with pandas `0.24.2` :(

**Comment by mrocklin** [comment]
> This fails tests because it is only compatible with pandas 0.24.2 :(

```python
if pandas.__version__ >= '0.24.2':
    return fancy_repr()
else:
    return old_repr()
```

?

**Comment by kkraus14** [comment]
How are we handling a situation like:
```
cudf.Series([1.0, np.nan, None], dtype='float32', nan_as_null=False)
```

**Comment by thomcom** [comment]
> > This fails tests because it is only compatible with pandas 0.24.2 :(
> 
> ```python
> if pandas.__version__ >= '0.24.2':
>     return fancy_repr()
> else:
>     return old_repr()
> ```
> 
> ?

Updating the tests to handle the new repr was a bear. I could do this and make the tests conditional also, however.

**Comment by thomcom** [comment]
> How are we handling a situation like:
> 
> ```
> cudf.Series([1.0, np.nan, None], dtype='float32', nan_as_null=False)
> ```
```
In [15]: cudf.Series([1.0, np.nan, None], dtype='float32', nan_as_null=False)
    ...:
Out[15]:
0    1.0
1    NaN
2    NaN
dtype: float32
```
Internally [1] is still `np.nan` and [2] is still `null`. Because Pandas doesn't distinguish between them they appear equal in the `print` output.

**Comment by thomcom** [comment]
This is superseded by https://github.com/rapidsai/cudf/pull/2432

**Pull request closed without merging**